rtl/eclrtl/rtlnewkey.cpp

```diff
@@ -1207,6 +1254,201 @@ IFieldFilter * createFieldFilter(unsigned fieldId, const RtlTypeInfo & type, con
+ */
+class SharedRtlTypeInfo : public CInterface
+{
+public:
+    SharedRtlTypeInfo(const RtlTypeInfo * _type) : type(_type) {}
+    ~SharedRtlTypeInfo() { type->doDelete(); }
+
+    const RtlTypeInfo * const type;
+};
+
+/*
+ * A field filter that can match a set of values.
+ */
+class SubStringFieldFilter : public SetFieldFilter
+{
+public:
+    SubStringFieldFilter(unsigned _field, const RtlTypeInfo & _fieldType, SharedRtlTypeInfo * _subType, IValueSet * _values)
+    : SetFieldFilter(_field, _fieldType, _values), subType(_subType)
+    {
+        subLength = subType->type->length;
+    }
+
+    virtual StringBuffer & serialize(StringBuffer & out) const override;
+    virtual MemoryBuffer & serialize(MemoryBuffer & out) const override;
+protected:
+    Linked<SharedRtlTypeInfo> subType;
+    size32_t subLength;
+};
+
+StringBuffer & SubStringFieldFilter::serialize(StringBuffer & out) const
+{
+    out.append(':').append(subLength).append("=");
+    return values->serialize(out);
+}
+
+MemoryBuffer & SubStringFieldFilter::serialize(MemoryBuffer & out) const
+{
+    out.append(':').append(subLength);
+    return values->serialize(out);
+}
+
+
+class FixedSubStringFieldFilter : public SubStringFieldFilter
+{
+public:
+    FixedSubStringFieldFilter(unsigned _field, const RtlTypeInfo & _fieldType, SharedRtlTypeInfo * _subType, IValueSet * _values)
+    : SubStringFieldFilter(_field, _fieldType, _subType, _values)
+    {
+    }
+
+    virtual IFieldFilter *remap(unsigned newField) const override { return new SubStringFieldFilter(newField, type, subType, values); }
+};
+
+
+
+class VariableSubStringFieldFilter : public SubStringFieldFilter
+{
+public:
+    VariableSubStringFieldFilter(unsigned _field, const RtlTypeInfo & _fieldType, SharedRtlTypeInfo * _subType, IValueSet * _values)
+    : SubStringFieldFilter(_field, _fieldType, _subType, _values),
+      maxTempLength(subLength * 4) //Maximum expansion from length to size is 4 bytes for utf8
+    {
+    }
+
+//Simple row matching
+    virtual bool matches(const RtlRow & row) const override;
+
+//More complex index matching
+    virtual int compareLowest(const RtlRow & left, unsigned range) const override;
+    virtual int compareHighest(const RtlRow & left, unsigned range) const override;
+    virtual int findForwardMatchRange(const RtlRow & row, unsigned & matchRange) const override;
+
+    virtual IFieldFilter *remap(unsigned newField) const override { return new VariableSubStringFieldFilter(newField, type, subType, values); }
+
+protected:
+    const unsigned maxTempLength;
+};
+
+
+bool VariableSubStringFieldFilter::matches(const RtlRow & row) const
+{
+    const byte * ptr = row.queryField(field);
+    size32_t len = *reinterpret_cast<const size32_t *>(ptr);
+    if (likely(len >= subLength))
+        return values->matches(ptr + sizeof(size32_t));
+
+    //Clone and expand the string to the expected length
+    byte * temp = (byte *)alloca(maxTempLength);
```

Comment:
What happens when people use STRING1000000 ... Do they get what they deserve?

Reply:
It would only by triggered for

    variablesizefieldstring[1..1000000] = 'x'

in which case, yes they probably get what they deserve!  I will add an assert that the substring length is < 1000.

Reply:
Looks like you added a check that less than 256 ...